runtime/vm/createramclass.cpp

```diff
@@ -1634,49 +1643,64 @@ loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9
 #if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
 /**
  * Attempts to pre-load classes of fields with Q signatures and no static
- * access modifier set.
+ * access modifier set. Also records largest field type in largestFieldType.
+ * 0 for single 1 for ref and 2 for double. Records total number of
+ * flattenable instance fields in flattenableInstanceFieldCount.
  *
  * Caller should not hold the classTableMutex.
  *
  * Return TRUE on success. On failure, returns FALSE and sets the
  * appropriate Java error on the VM.
  */
 static BOOLEAN
-loadFlattenableFieldValueClasses(J9VMThread *currentThread, J9ClassLoader *classLoader, J9ROMClass *romClass, UDATA classPreloadFlags, UDATA packageID, J9Module *module)
+loadFlattenableFieldValueClasses(J9VMThread *currentThread, J9ClassLoader *classLoader, J9ROMClass *romClass, UDATA classPreloadFlags, UDATA packageID, J9Module *module, UDATA *largestFieldType, UDATA *flattenableInstanceFieldCount)
 {
 	J9ROMFieldWalkState fieldWalkState = {0};
 	J9ROMFieldShape *field = romFieldsStartDo(romClass, &fieldWalkState);
 	BOOLEAN result = TRUE;
 	/* iterate over fields and load classes of fields marked as QTypes */
 	while (NULL != field) {
 		const U_32 modifiers = field->modifiers;
-		J9UTF8 *signature = J9ROMFIELDSHAPE_SIGNATURE(field);
-		U_8 *signatureChars = J9UTF8_DATA(signature);
-		if (('Q' == signatureChars[0]) && J9_ARE_NO_BITS_SET(modifiers, J9AccStatic)) {
-			J9Class *valueClass = internalFindClassUTF8(currentThread, signatureChars + 1, J9UTF8_LENGTH(signature) - 2, classLoader, classPreloadFlags);
-			if (NULL == valueClass) {
-				result = FALSE;
-				goto done;
-			} else {
-				J9ROMClass *valueROMClass = valueClass->romClass;
-
-				if (J9_ARE_NO_BITS_SET(valueROMClass->modifiers, J9AccValueType)) {
-					J9UTF8 *badClass = NNSRP_GET(valueROMClass->className, J9UTF8*);
-					setCurrentExceptionNLSWithArgs(currentThread, J9NLS_VM_ERROR_QTYPE_NOT_VALUE_TYPE, J9VMCONSTANTPOOL_JAVALANGINCOMPATIBLECLASSCHANGEERROR, J9UTF8_LENGTH(badClass), J9UTF8_DATA(badClass));
+		if (J9_ARE_NO_BITS_SET(modifiers, J9AccStatic)) {
+			J9UTF8 *signature = J9ROMFIELDSHAPE_SIGNATURE(field);
+			U_8 *signatureChars = J9UTF8_DATA(signature);
+			if (('Q' == signatureChars[0]) && J9_ARE_NO_BITS_SET(modifiers, J9AccStatic)) {
```

Comment:
Extra static check here. Would this read better as a switch statement instead of elseif?